src/main/java/de/hysky/skyblocker/skyblock/dwarven/PowderMiningTracker.java

```diff
@@ -0,0 +1,339 @@
+	 * Once the filter is changed, the {@link #SHOWN_REWARDS} map is cleared and recalculated based on this map.
+	 * </p>
+	 * <p>This is similar to how {@link ChatHud#messages} and {@link ChatHud#visibleMessages} behave.</p>
+	 *
+	 * @implNote This is a map of item IDs to the amount of that item obtained.
+	 */
+	@SuppressWarnings("JavadocReference")
+	private static final Object2IntMap<String> ALL_REWARDS = new Object2IntArrayMap<>();
+	private static final Object2ObjectArrayMap<String, String> NAME2ID_MAP = new Object2ObjectArrayMap<>(50);
+	private static boolean insideChestMessage = false;
+	private static double profit = 0;
+
+	@SuppressWarnings("BooleanMethodIsAlwaysInverted")
+	private static boolean isEnabled() {
+		return SkyblockerConfigManager.get().mining.crystalHollows.enablePowderTracker;
+	}
+
+	@Init
+	public static void init() {
+		ChatEvents.RECEIVE_STRING.register(text -> {
+			if (Utils.getLocation() != Location.CRYSTAL_HOLLOWS || !isEnabled()) return;
+			// Reward messages end with a separator like so
+			if (insideChestMessage && text.equals("▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬")) {
+				insideChestMessage = false;
+				return;
+			}
+
+			if (!insideChestMessage && (text.equals("  CHEST LOCKPICKED ") || (SkyblockerConfigManager.get().mining.crystalHollows.countNaturalChestsInTracker && text.equals("  LOOT CHEST COLLECTED ")))) {
+				insideChestMessage = true;
+				return;
+			}
+
+			if (!insideChestMessage) return;
+			Matcher matcher = REWARD_PATTERN.matcher(text);
+			if (!matcher.matches()) return;
+			String itemName = matcher.group(1);
+			int amount = NumberUtils.toInt(matcher.group(2).replace(",", ""), 1);
+
+			String itemId = getItemId(itemName);
+			if (itemId.isEmpty()) {
+				LOGGER.error("No matching item id for name `{}`. Report this!", itemName);
+				return;
+			}
+			incrementReward(itemName, itemId, amount);
+			calculateProfitForItem(itemId, amount);
+		});
+
+		HudRenderEvents.AFTER_MAIN_HUD.register((context, tickCounter) -> {
+			if (Utils.getLocation() != Location.CRYSTAL_HOLLOWS || !isEnabled()) return;
+			int y = MinecraftClient.getInstance().getWindow().getScaledHeight() / 2 - 100;
+			var set = SHOWN_REWARDS.object2IntEntrySet();
+			for (Object2IntMap.Entry<Text> entry : set) {
+				context.drawTextWithShadow(MinecraftClient.getInstance().textRenderer, entry.getKey(), 5, y, 0xFFFFFF);
+				context.drawTextWithShadow(MinecraftClient.getInstance().textRenderer, Text.of(String.valueOf(entry.getIntValue())), 10 + MinecraftClient.getInstance().textRenderer.getWidth(entry.getKey()), y, 0xFFFFFF);
+				y += 10;
+			}
+			context.drawTextWithShadow(MinecraftClient.getInstance().textRenderer, Text.literal("Gain: " + NumberFormat.getInstance().format(profit) + " coins").formatted(Formatting.GOLD), 5, y + 10, 0xFFFFFF);
+		});
+
```

Comment:
I would move the hud rendering and chat message logic to new methods so that the `init` stays concise